### PR TITLE
bug: Tighten type constraints on DenseArrayBase.create_dense_int_or_index

### LINKED ISSUE #3154: Function _parse_builtin_densearray_attr fails when parsing a dense array with element_type IndexType
The function `_parse_builtin_densearray_attr` does not accept dense arrays with element_type `IndexType`. On the other hand, the function `DenseArrayBase.create_dense_int_or_index` does accept `IndexType` as element_type.

### PR DESCRIPTION
**Pull request by vinayakdsci**
Fixes #3154.

MLIR supports the following data types for `DenseArrayAttr`: `bool`, `i8`, `i16`, `i32`, `i64`, `f32`, and `f64`.

However, we also supported `index` in DenseArrayAttrs through the function `DenseArrayBase.create_dense_int_or_index()`, which caused an incompatibility with MLIR, causing `mlir-opt` to crash on IR produced by xdsl in this context.

This patch renames the function `DenseArrayBase.create_dense_int_or_index()` to `DenseArrayBase.create_dense_int()` so that the function's name is not misleading, and tightens type constraints on the function's inputs to only be of IntegerType.

This removes the incompatibility with MLIR produced from earlier behavior.

### CODE REVIEW

**Comment by vinayakdsci** [comment]
cc @superlopuh @watermelonwolverine.

**Comment by superlopuh** [comment]
Could you please add back the `create_int_or_index` function with a deprecated warning, advising users to use the new one instead? We can delete it later, but clients that depend on xDSL have one less breaking change

File: xdsl/dialects/builtin.py [hunk 1]
```diff
@@ -941,6 +941,18 @@ def verify(self):
...
+    @staticmethod
+    def create_dense_int_or_index(
+        data_type: IntegerType | IndexType, data: Sequence[int] | Sequence[IntAttr]
+    ) -> DenseArrayBase:
+        if len(data) and isinstance(data[0], int):
+            attr_list = [IntAttr(d) for d in cast(Sequence[int], data)]
+        else:
+            attr_list = cast(Sequence[IntAttr], data)
+
+        return DenseArrayBase([data_type, ArrayAttr(attr_list)])
```
Comment (superlopuh) [inline xdsl/dialects/builtin.py:954]:
```suggestion
        assert not isinstance(data_type, IndexType), "Index type is not supported"
        return DenseArrayBase.create_int(data_type, data)
```

**Comment by vinayakdsci** [comment]
@math-fehr @superlopuh does this PR look good to go?

**Review by math-fehr — APPROVED** [review]
Yes, sorry I missed the previous pings!

**Comment by vinayakdsci** [comment]
@math-fehr tests have passed, can we merge the PR?

Thanks a lot!

**Comment by math-fehr** [comment]
> @math-fehr tests have passed, can we merge the PR?
> 
> 
> 
> Thanks a lot!

Sorry I got distracted when updating it, merging it now, thanks a lot!

**Pull request merged**